packages/common/src/types/filter.ts

```diff
@@ -199,6 +199,60 @@ export const updateFieldIdInFilters = (
     }
 };
 
+export const isMetricToDelete = (
+    item: FilterGroupItem,
+    metricName: string,
+): boolean => !isFilterGroup(item) && item.target.fieldId === metricName;
+
+export const removeMetricFromFilterGroupItem = (
+    item: FilterGroupItem,
+    metricName: string,
+): void => {
+    if (isFilterGroup(item)) {
+        // eslint-disable-next-line @typescript-eslint/no-use-before-define
+        removeMetricFromFilters(item, metricName);
+    }
+};
+
+export const removeMetricFromFilters = (
+    filterGroup: FilterGroup | undefined,
+    metricName: string,
+): void => {
+    if (!filterGroup) return;
+
+    const processGroupItems = (items: FilterGroupItem[]): FilterGroupItem[] =>
+        items
+            .filter((item) => !isMetricToDelete(item, metricName))
+            .map((item) => {
+                removeMetricFromFilterGroupItem(item, metricName);
+                return item;
+            });
+
+    const isNotEmptyGroup = (item: FilterGroupItem): boolean => {
+        if (isOrFilterGroup(item)) return item.or.length !== 0;
+        if (isAndFilterGroup(item)) return item.and.length !== 0;
+        return true;
+    };
+
+    /* eslint-disable no-param-reassign */
+    if (isOrFilterGroup(filterGroup)) {
+        filterGroup.or = processGroupItems(filterGroup.or);
+        if (filterGroup.or.length === 0) filterGroup = undefined;
+        else
+            filterGroup.or = filterGroup.or.filter((item) =>
+                isNotEmptyGroup(item),
+            );
```

Comment:
Since this is a recursive function, do we need this `else`? wouldn't the previous line `if (filterGroup.or.length === 0) filterGroup = undefined;` remove all empty groups?

Reply:
@ZeRego 
This case was mostly to filter out the cases when the `and`, `or` are not of length 0 but rather contain a empty nested `and`, `or` inside. I realized this might fail while writing test cases due to the way the parameters are updated locally inside a function. There would be problems setting it to undefined as it is passed as a reference. So I have moved the code for undefined separately.
It works fine now. Just wanted to know, do you think it shouldn't be this complicated?

Reply:
> Just wanted to know, do you think it shouldn't be this complicated?

Thanks for the changes and especially the tests ❤️  I still ended up refactoring the function as I felt it could be simplified even more. 
It got easier to read and test once I removed the mutability nature. Then I was able to consolidate all the functions into 1 and only check once for empty groups.